etcdserver/api/v3rpc/watch.go

```diff
@@ -190,17 +194,32 @@ func (ws *watchServer) Watch(stream pb.Watch_WatchServer) (err error) {
 		close(sws.ctrlStream)
 
 	case <-stream.Context().Done():
-		err = stream.Context().Err()
-		// the only server-side cancellation is noleader for now.
-		if err == context.Canceled {
-			err = rpctypes.ErrGRPCNoLeader
-		}
+		err = getClientStreamClosureError(stream.Context(), sws.sg)
 	}
 
 	sws.close()
 	return err
 }
 
+// getClientStreamClosureError tries to determine a more specific client stream
+// cancellation error. If the context error appears to be a cancellation, use
+// the stream context and local leader state to detect leader loss. If the
+// reason is inconclusive, assume a client cancellation.
+func getClientStreamClosureError(ctx context.Context, statusGetter etcdserver.RaftStatusGetter) error {
+	err := ctx.Err()
+	if err == context.Canceled {
+		err = rpctypes.ErrGRPCWatchCanceled
+		if md, hasMetadata := metadata.FromIncomingContext(ctx); hasMetadata {
+			if rl := md[rpctypes.MetadataRequireLeaderKey]; len(rl) > 0 && rl[0] == rpctypes.MetadataHasLeader {
+				if statusGetter.Leader() == types.ID(raft.None) {
+					err = rpctypes.ErrGRPCNoLeader
```

Comment:
Does anyone have an idea how to induce this state in an e2e test?

Reply:
I recently enabled metrics reporting and stumbled upon this, too.  I'm not able to reproduce it at will, though it does frequently.